src/unix/stream.c

```diff
@@ -795,7 +799,7 @@ static void uv__write(uv_stream_t* stream) {
     msg.msg_iovlen = iovcnt;
     msg.msg_flags = 0;
 
-    msg.msg_control = (void*) scratch;
+    msg.msg_control = (void*) &scratch.alias;
```

Comment:
I tried your suggestion. It fails with the same symptoms as before :-(.
I think this is what is happening.

According to strict aliasing,  `struct cmsghdr *cmsg` and `&scratch[0]` will never alias each other because they point to different types. So the optimizer is free to re-order operations on them. 

msg.msg_control = **&scratch[0];**
...
cmsg = CMSG_FIRSTHDR(&msg);
**cmsg->cmsg_level** = SOL_SOCKET;

But they do point to the same memory when using `msg.msg_control = &scratch[0];`  

The union seems to prevent the re-ordering.
```
union {
      char data[64];
      char alias;
    } scratch;
```

I now realize that it might be passing by fluke. A correct union would be this. (which also passes).
```
union {
      char data[64];
      struct cmsghdr alias;
    } scratch;
```